airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py

```diff
@@ -284,16 +284,18 @@ def _read_full_refresh(
                 if self._limit_reached(internal_config, total_records_counter):
                     return
 
-    @staticmethod
-    def _checkpoint_state(stream: Stream, stream_state, state_manager):
+    def _checkpoint_state(self, stream: Stream, stream_state, state_manager: ConnectorStateManager):
         # First attempt to retrieve the current state using the stream's state property. We receive an AttributeError if the state
         # property is not implemented by the stream instance and as a fallback, use the stream_state retrieved from the stream
         # instance's deprecated get_updated_state() method.
         try:
             state_manager.update_state_for_stream(stream.name, stream.namespace, stream.state)
         except AttributeError:
             state_manager.update_state_for_stream(stream.name, stream.namespace, stream_state)
-        return AirbyteMessage(type=MessageType.STATE, state=AirbyteStateMessage(data=state_manager.get_legacy_state()))
+        if self.per_stream_state_enabled:
+            return state_manager.create_state_message(stream.name, stream.namespace)
+        else:
+            return AirbyteMessage(type=MessageType.STATE, state=AirbyteStateMessage(data=dict(state_manager.get_legacy_state())))
```

Comment:
i like this change, this will also allow us to streamline the interface and `_get_legacy_state()` can become a method that is internally used by the state manager.